### PR TITLE
fix credentials matching "by realm" / improve WWW-Authenticate: header parsing

### PR DESCRIPTION
**Pull request by cchepelov**
With certain repository managers, such as recent versions of the Sonatype Nexus Repository Manager, the WWW-Authenticate header line returned during an HTTP/401 result contains the desired realm, but also additional fields.

In case the authentication credentials are supplied by sbt, the "authentication" is always empty, until an http/401 bounce enables Coursier to locate the correct credentials.

With repositories behaving as above, this causes Coursier to fail to match the realm with the list of user-supplied credentials, and ultimately builds fail by never being able to access authentication-protected artifacts.

This PR revises the parsing of the WWW-Authenticate line by making it more tolerant to server-supplied variations 

Fixes: #1421

### CODE REVIEW

**Comment by cchepelov** [comment]
Taking the liberty to summarily dismiss Mima: we're changing the interface of a `private[coursier] class`

**Comment by cchepelov** [comment]
Note: I haven't tested, but I suspect this also fixes a likely similar issue when using the CLI with credentials supplied per https://get-coursier.io/docs/other-credentials when (i) relying on realms matching and (ii) the repositories exhibit the same "extraneous fields" behaviour.

#TIL that [RFC  7617](https://tools.ietf.org/html/rfc7617#section-2.1) supersedes RFC 2617 and indeed provides for the ", charset="UTF-8" thing.
And yes, the only allowed charset is UTF-8.

**Review by alexarchambault — COMMENTED** [review]
Thanks @cchepelov, that looks fine!

Maybe we could add a non-regression test for it in [this file](https://github.com/coursier/coursier/blob/2a18d02a5f5eb26900bd7a647cfd98d222f7eb6d/modules/cache/jvm/src/test/scala-2.12/coursier/cache/FileCacheTests.scala)? There are examples of tests spawning an HTTP server from an `org.http4s.HttpService` (look for `withHttpServer`), then checking that `FileCache` can fetch things fine from it, optionally passing credentials to it (look for `expect` and `addCredentials`). `HttpService` would allow to answer requests with custom HTTP headers, like a `WWW-Authenticate` with a `charset="UTF-8"` for us here.

File: modules/cache/jvm/src/main/scala/coursier/cache/CacheUrl.scala [hunk 1]
```diff
@@ -387,14 +388,49 @@ object CacheUrl {
...
+
+    def unapply(wwwAuthenticate: String): Option[String] = {
+      wwwAuthenticate match {
+        case BasicAuthBase(basicAuthLine) =>
+          Param.findAllMatchIn(basicAuthLine).map { mobj =>
+            val key = mobj.group(1)
+            val value = mobj.group(2)
+            println(s"in wwwAuthenticate line ${wwwAuthenticate}, found (${key}->${value})")
+            mobj.group(1).toLowerCase(Locale.ROOT) -> mobj.group(2)
+          }.collect { case ("realm", realm) => realm }.find(_ => true)
```
Comment (alexarchambault) [inline modules/cache/jvm/src/main/scala/coursier/cache/CacheUrl.scala:427]: Change this to
```suggestion
          Param.findAllMatchIn(basicAuthLine)
            .map(mobj => mobj.group(1).toLowerCase(Locale.ROOT) -> mobj.group(2))
            .collectFirst { case ("realm", realm) => realm }
```
?

Comment (cchepelov) [reply modules/cache/jvm/src/main/scala/coursier/cache/CacheUrl.scala:427]: DUH, of course !
(furthermore, well-spotted, I forgot to remove the println)

File: project/Mima.scala [hunk 2]
```diff
@@ -1,4 +1,3 @@
-
```
[the remark this answers is not in the document]
Comment (cchepelov) [reply project/Mima.scala:1]: whoops, that's entirely accidental, didn't realise it even happened. Sorry.

[the remark this answers is not in the document]
Comment (cchepelov) [reply project/Mima.scala]: (redid the change & rebased away the problem)

**Comment by cchepelov** [comment]
re. non-reg: aaah, yes! Didn't realise there _was_ coverage there (initially wanted to solve that by red-green'ing it).

Let me come back to you

**Comment by cchepelov** [comment]
done. And that wasn't in vain: I noticed the regex-based parser was a bit too lenient (after the first commit, it would have passed 
`www-authenticate: Basic ç&'à&'&lzejfz param="toto" qlsdfklqmdghfzkehigfzmelihfzl realm="ABCD" $$$$ foo="bar"`

**Review by alexarchambault — APPROVED** [review]
Thanks!

**Pull request merged**